mockito-core/src/main/java/org/mockito/internal/PremainAttachAccess.java

```diff
@@ -34,16 +40,41 @@ public static Instrumentation getInstrumentation() {
         }
         if (instrumentation == null) {
             if (ClassFileVersion.ofThisVm().isAtLeast(ClassFileVersion.JAVA_V21)) {
-                // Cannot use `Plugins.getMockitoLogger().warn(...)` at this time due to a circular
-                // dependency on `Plugins.registry`.
-                // The `PluginRegistry` is not yet fully initialized (in `Plugins`), because it is
-                // currently initializing the `MockMaker` which is a InlineByteBuddyMockMaker, and
-                // it is later calling this method to access the instrumentation.
-                System.err.println(
-                        "Mockito is currently self-attaching to enable the inline-mock-maker. This "
-                                + "will no longer work in future releases of the JDK. Please add Mockito as an agent to your "
-                                + "build as described in Mockito's documentation: "
-                                + "https://javadoc.io/doc/org.mockito/mockito-core/latest/org/mockito/Mockito.html#0.3");
+                boolean dynamicAgentLoading;
+                try {
+                    Object runtimeMXBean =
+                            Class.forName("java.lang.management.ManagementFactory")
+                                    .getMethod("getRuntimeMXBean")
+                                    .invoke(null);
+                    @SuppressWarnings("unchecked")
+                    List<String> arguments =
+                            (List<String>)
+                                    runtimeMXBean
+                                            .getClass()
+                                            .getMethod("getInputArguments")
+                                            .invoke(runtimeMXBean);
+                    dynamicAgentLoading =
+                            arguments.stream()
+                                    .anyMatch(
+                                            argument ->
+                                                    argument.contains(
+                                                            "-XX:+EnableDynamicAgentLoading"));
+                } catch (Exception ignored) {
```

Comment:
Can we update the tests we have for this feature to ensure this as well? If I remember correctly we had those tests in the `.gradle` files somewhere

Reply:
Fair point, I tried to add a test!